internal/common/expiring_cache.go

```diff
@@ -0,0 +1,46 @@
+package common
+
+import (
+	"sync"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+)
+
+type ExpiringCache interface {
+	GetOrInsert(key string, value interface{}) (interface{}, bool)
+	Get(key string) (interface{}, bool)
+	Set(key string, value interface{})
+}
+
+type expiringCache struct {
+	cache *cache.Cache
+	mutex sync.Mutex
+}
+
+func NewExpiringCache(defaultExpiration, cleanupInterval time.Duration) ExpiringCache {
+	return &expiringCache{cache: cache.New(defaultExpiration, cleanupInterval)}
+}
+
+func (e *expiringCache) GetOrInsert(key string, value interface{}) (interface{}, bool) {
```

Comment:
The GetOrInsert is done in atomic way.  If the value already exists it returns the existing value.  Otherwise it adds and returns the new one.
We do not want a case that 2 threads at the same time will attempt to insert a value for a specific key and one will overwrite the other.